ios-base/Navigators/Navigator.swift

```diff
@@ -0,0 +1,204 @@
+  open var currentViewController: UIViewController?
+
+  public required init(with route: Route) {
+    let screen = route.screen
+    if let navigation = screen as? UINavigationController {
+      rootViewController = navigation
+    } else {
+      let viewController = screen
+      rootViewController = UINavigationController(rootViewController: viewController)
+      rootViewController?.isNavigationBarHidden = true
+      currentViewController = viewController
+    }
+  }
+}
+
+/**
+ The Navigator is the base of this framework, it handles all
+ your application navigation stack.
+ It keeps track of your root NavigationController and the
+ ViewController that is currently displayed. This way it can
+ handle any kind of navigation action that you might want to dispatch.
+ */
+public protocol Navigator: class {
+  /// The root navigation controller of your stack.
+  var rootViewController: UINavigationController? { get set }
+
+  /// The currently visible ViewController
+  var currentViewController: UIViewController? { get set }
+
+  /// Convencience init to set your application starting screen.
+  init(with route: Route)
+
+  /**
+   Navigate from your current screen to a new route.
+   - Parameters:
+   - route: The destination route of your navigation action.
+   - transition: The transition type that you want to use.
+   - animated: Animate the transition or not.
+   - completion: Completion handler.
+   */
+  func navigate(to route: Route, with transition: TransitionType, animated: Bool, completion: (() -> Void)?)
+
+  /**
+   Navigate from your current screen to a new entire navigator. Can only push a router as a modal. Afterwards, other controllers can be pushed inside the presented Navigator.
+   - Parameters:
+   - Navigator: The destination navigator that you want to navigate to
+   - animated: Animate the transition or not.
+   - completion: Completion handler.
+   */
+  func navigate(to router: Navigator, animated: Bool, completion: (() -> Void)?)
+
+  /**
+   Handles backwards navigation through the stack.
+   */
+  func pop(animated: Bool)
+
+  /**
+   Handles backwards navigation through the stack.
+   - Parameters:
+   - route: The index of the route of your navigation action.
+   - animated: Animate the transition or not.
+   */
+  func popTo(index: Int, animated: Bool)
+
+  /**
+   Handles backwards navigation through the stack.
+   - Parameters:
+   - route: The destination route of your navigation action.
+   - animated: Animate the transition or not.
+   */
+  func popTo(route: Route, animated: Bool)
+
+  /**
+   Dismiss your current ViewController.
+   - Parameters:
+   - animated: Animate the transition or not.
+   - completion: Completion handler.
+   */
+  func dismiss(animated: Bool, completion: (() -> Void)?)
+}
+
+public extension Navigator {
+  func navigate(to route: Route, with transition: TransitionType, animated: Bool = true, completion: (() -> Void)? = nil) {
```

Comment:
how would you handle custom transitions ?

Reply:
We need to add a new transitionType, I was thinking about doing something like `TransitonType.custom` and that one would have a configuration callback that receives the presenting and presented viewControllers as parameters, so you can configure delegates, presentationStyles, etc.

Reply:
Ended up doing something like this, but at `Route` level, I think it makes more sense, since the view is the one who knows the actual views.